src/feature.js

```diff
@@ -123,103 +131,88 @@ class Feature {
-    // Let the content know we've navigated.
-    browser.webNavigation.onCompleted.addListener(() => {
-      if (feature.portFromCS) {
-        feature.portFromCS.postMessage({message: "navigation"});
+    // Whenever trackers are detected on a tab, record their presence.
+    browser.trackers.listenForTrackers();
+    browser.trackers.onRecordTrackers.addListener(
+      tabId => {
+        if (tabId < 0) {
+          return;
+        }
+        let tabInfo = TabRecords.getOrInsertTabInfo(tabId);
+        tabInfo.hasTrackers = true;
       }
-    });
+    );
 
-    browser.trackers.listenForTrackers();
+    // When the top-level location of a tab changes, submit telemetry probes
+    // for the old page and reset the payload to record the new page.
     browser.trackers.onLocationChanged.addListener(
-      () => {
-        // reset tracker status when location changes
-        this.trackersOnPage = false;
+      async (tabId) => {
+        if (tabId < 0) {
+          return;
+        }
+        let tabInfo = TabRecords.getOrInsertTabInfo(tabId);
+        // Only submit telemetry if we have recorded load info and
+        // the tab actually has trackers.
+        if (tabInfo.telemetryPayload.etld && tabInfo.hasTrackers) {
+          this.sendTelemetry(tabInfo.telemetryPayload);
+        }
+        TabRecords.resetPayload(tabId);
+        tabInfo.hasTrackers = false;
       }
     );
 
-    browser.trackers.onRecordTrackers.addListener(
-      () => {
-        this.trackersOnPage = true;
+    // When a tab is removed, make sure to submit telemetry for the
+    // last page and delete the tab entry.
+    browser.tabs.onRemoved.addListener(tabId => {
+      let tabInfo = TabRecords.getOrInsertTabInfo(tabId);
+      // Only submit telemetry if we have recorded load info and
+      // the tab actually has trackers.
+      if (tabInfo.telemetryPayload.etld && tabInfo.hasTrackers) {
+        this.sendTelemetry(tabInfo.telemetryPayload);
       }
-    );
+      TabRecords.deleteTabEntry(tabId);
+    });
 
-    browser.notificationBar.onSurveyShown.addListener(
-      () => {
-        feature.addToPayload({
-          SURVEY_SHOWN: "true",
-        });
-      },
-    );
+    // Watch for messages from the content script about page information
+    // such as whether the page was reloaded and load timing info.
+    browser.runtime.onConnect.addListener((p) => {
+      p.onMessage.addListener(({data}) => {
+        let tabInfo = TabRecords.getOrInsertTabInfo(p.sender.tab.id);
+        tabInfo.telemetryPayload.etld = this.SHA256(userid + data.etld);
+        tabInfo.telemetryPayload.page_reloaded = data.pageReloaded;
+        for (let key in data.performanceEvents) {
+          tabInfo.telemetryPayload[key] = data.performanceEvents[key];
+        }
 
+        // Show the user a survey if the page was reloaded.
+        if (data.pageReloaded) {
+          tabInfo.reloadCount += 1;
+          if (tabInfo.hasTrackers) {
```

Comment:
nit: combine with the `if` above, no point in incrementing the counter or recording anything at all. We could also just return first thing in the listener if no trackers?

Reply:
Since the tracker information arrives asynchronously I'd prefer to record the payload anyway. Right now the content script message arrives on "load" which means that tracker information is guaranteed to arrive beforehand, but there's a reasonable argument to make that we might want to send data at an earlier point already...